### PR TITLE
rewrite main gui to be non-destructive and faster

### PR DESCRIPTION
**Pull request by XVhc6A**
### Brief description of the changes:
Instead of tearing down the gui and then building it from scratch, it is relatively simple to update the values of each element. (we can use .visible instead of .destroy() to get rid of things)

I think the last place to squeeze some performance out is to somehow create a single gui once for everyone, and then only set .visible flags per person in the update_main_gui function. but I think I will give up here since I have wasted enough time on this.

this touches a lot of code but with a single person in lobby the average time for .refresh has gone from about 4ms to 0.8ms... but with maximum spikes changing from 10.6 ms to 3ms

A truly laughable improvement.... :'(

but maybe it will be more improved with more people.....

anyway, I didn't test this in multiplayer or anything so there's probably little bits that are broken - I'm not super well versed in what causes desyncs.

I am also surprised at how long reanimate_units takes. I wonder if that should be moved into main.lua or something because it's a bit of a resource hog even at 0 evo.

also worth noting, it can be very hard to profile code because of the on_generate_chunk code... it seems that is not very friendly.....

also the difficulty option meter moved... can be easily fixed but I am taking a break from this code.

### Tested Changes:
- [ ] I've tested the changes locally or with people.
- [ ] I've not tested the changes.

### CODE REVIEW

File: maps/biter_battles_v2/gui.lua [hunk 1]
```diff
@@ -44,70 +44,191 @@ local function create_sprite_button(player)
 	element_style({element = button, x= 38, y = 38, pad = -2})
 end
 
-local function clock(frame)
+local function get_current_clock_time_string()
 	local total_minutes = math.floor(game.ticks_played / (60 * 60))
 	local total_hours = math.floor(total_minutes / 60)
 	local minutes = total_minutes - (total_hours * 60)
+	return "Game time: " .. string.format("%02d", total_hours) .. ":" .. string.format("%02d", minutes)
```
Comment (AwesomePatrol) [inline maps/biter_battles_v2/gui.lua:51]: `table.concat` should be used instead for performance reasons (successive concatenations need to copy the string as strings are immutable in lua)

File: maps/biter_battles_v2/gui.lua [hunk 2]
```diff
@@ -357,17 +489,18 @@ function join_team(player, force_name, forced_join, auto_join)
...
 	local i = player.get_inventory(defines.inventory.character_main)
 	i.clear()
-	player.insert {name = 'pistol', count = 1}
-	player.insert {name = 'raw-fish', count = 3}
-	player.insert {name = 'firearm-magazine', count = 32}
-	player.insert {name = 'iron-gear-wheel', count = 8}
-	player.insert {name = 'iron-plate', count = 16}
-	player.insert {name = 'burner-mining-drill', count = 10}
-	player.insert {name = 'wood', count = 2}
+	player.insert{name = 'pistol', count = 1}
```
Comment (AwesomePatrol) [inline maps/biter_battles_v2/gui.lua:492]: Please avoid changes in other parts of the code if possible

File: maps/biter_battles_v2/gui.lua [hunk 3]
```diff
@@ -131,135 +253,147 @@ local function add_prod_button(elem, gui_value)
...
 		if not first_team then
-			frame.add { type = "line", caption = "this line", direction = "horizontal" }
+			frame.add{ type = "line", direction = "horizontal" }
 		else
 			first_team = false
 		end
 
 		-- Team name & Player count
-		local t = frame.add { type = "table", column_count = 4 }
+		local t = frame.add{ name="table", type = "table", column_count = 4 }
```
Comment (AwesomePatrol) [inline maps/biter_battles_v2/gui.lua:350]: Could you please use more unique name?

**Review by AwesomePatrol — COMMENTED** [review]
This looks promising. With small cleanup we can get to testing in multiplayer

**Comment by AwesomePatrol** [comment]
There are some issues left (missing science button in spectate):
```
  51.829 Script @/home/patrol/.factorio/temp/currently-playing/utils/event_core.lua:48: LuaGuiElement API call when LuaGuiElement was invalid.
  55.275 Script @/home/patrol/.factorio/temp/currently-playing/utils/event_core.lua:48: ...rio/temp/currently-playing/maps/biter_battles_v2/gui.lua:281: attempt to index field 'bb_science_frame' (a nil value)
  65.625 Script @/home/patrol/.factorio/temp/currently-playing/utils/event_core.lua:48: LuaGuiElement API call when LuaGuiElement was invalid.
  ```

**Comment by XVhc6A** [comment]
fixed that @AwesomePatrol

**Comment by developer-8** [comment]
After changing the difficulty, the difficulty button does not change

**Comment by Ragnarok77-factorio** [comment]
for @XVhc6A  : 

=>I opened gui with fish, closed it, had this error : 
utils/event_core.lua:48: LuaGuiElement API call when LuaGuiElement was invalid

=>When I host it, I see the buttons in black instead of previous color (harder to read now), example here of the screen before and after your change : 
image

=>When I clicked auto join, got this error : 
/event_core.lua:48: ...rio/temp/currently-playing/maps/biter_battles_v2/gui.lua:513: attempt to index field 'bb_main_gui' (a nil value)

=>When I feed red science in training mode, no threat is increased/shown on gui, or even not training mode, it shows 0 (maybe just an illusion)

**Comment by Ragnarok77-factorio** [comment]
Feel free to open it again ! Don't hesitate at all

**Pull request closed without merging**